### DISCUSSION TITLE
Idea: Extension support

### DISCUSSION THREAD

**Discussion by DerFuchs** [JellyfishSDK/jellyfish#1540] [Ideas]
Hey there!

Since Christian Sandrini showed up his [Pond project](https://www.reddit.com/r/defiblockchain/comments/v7x6te/cfp_pond_decentralize_ocean_api_and_run_it/) I got an idea.

When I am able to host an Ocean instance by myself by using Pond, it'd be awesome to have the ability to extend Oceans functionality and endpoints via some kind of plug-ins.

I'd like to have the ability to add more custom endpoints to the system which allow me to request data more efficiently or even request data which is not available right now.

Quick example: Masternode Monitor basically shows which blocks have been minted by a master node. It's not possible to receive this information directly neither from a node itself, nor from Ocean. My current solution was to develop my own API which will then be able to fish this data out of the blockchain by processing all necessary data and providing a list of minted blocks. 

I believe it would be way better to provide an extension for Ocean which is able to deliver this list directly. Then I could set up a Pond instance for MaMon, integrate that extension and then there will be no more need for a custom API and MaMon could receive all it's data from a single source. Users will also be able to switch to their own custom data source by running a Pond instance with the necessary extension.

What are your thoughts about that?

**Comment by fuxingloh** [comment]
We have a "plugin" design for jellyfish and we call it [`jellyfish/apps`](https://github.com/JellyfishSDK/jellyfish/tree/main/apps). Our 2nd most used app is [`apps/playground-api`](https://github.com/orgs/JellyfishSDK/packages?repo_name=jellyfish) for testing. It provides additional functionality in that manner. We also have [`apps/status-api`](https://github.com/JellyfishSDK/jellyfish/pkgs/container/status-api) contributed by our ecosystem team. 2 other WIP ecosystems `apps/` were underway but I think the team that was working on them are busy with other things now.

Instead of building it directly into the `ocean-api/whale-api` stack, those are stand-alone systems that rely on `ocean-api` and build their own indexes for the very similar purpose you described above. You can use that very similar pattern to design your own public/private app!
